File: android/secure-core-android/src/main/kotlin/com/securecore/export/ExportService.kt
package com.securecore.export

import android.net.Uri
import com.securecore.SecureCoreLib
import com.securecore.SecureCoreResult
import com.securecore.keymanager.KeyManager
import com.securecore.metadata.DocumentEntity
import com.securecore.metadata.MetadataRepository
import com.securecore.store.DocumentStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.OutputStream
import java.security.MessageDigest
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import java.util.zip.ZipOutputStream

data class ExportReport(
    val exportedCount: Int,
    val failedCount: Int,
    val failedDocIds: List<String>
)

data class ImportReport(
    val importedCount: Int,
    val skippedCount: Int,
    val failedCount: Int,
    val failedDocIds: List<String>
)

/**
 * Handles export and import of recovery bundles.
 *
 * See docs/recovery-format-v1.md for the bundle format specification.
 */
class ExportService(
    private val keyManager: KeyManager,
    private val documentStore: DocumentStore,
    private val metadataRepository: MetadataRepository,
    private val outputStreamProvider: (Uri) -> OutputStream,
    private val inputStreamProvider: (Uri) -> java.io.InputStream
) {

    /**
     * Exports selected documents into a recovery bundle zip.
     *
     * For each document:
     * 1. Unwraps the DEK from the device keystore
     * 2. Re-wraps the DEK with the passphrase via Rust FFI
     * 3. Writes the .enc file, metadata, and recovery wrap to the zip
     * 4. Zeroizes the DEK
     */
    suspend fun exportBundle(
        docIds: List<String>,
        passphrase: String,
        outputUri: Uri
    ): Result<ExportReport> = withContext(Dispatchers.IO) {
        val failed = mutableListOf<String>()
        var exported = 0

        try {
            val outputStream = outputStreamProvider(outputUri)
            ZipOutputStream(outputStream).use { zip ->
                for (docId in docIds) {
                    try {
                        exportDocument(zip, docId, passphrase)
                        exported++
                    } catch (e: Exception) {
                        failed.add(docId)
                    }
                }

                // Write manifest
                val manifest = buildManifest(docIds.filter { it !in failed })
                zip.putNextEntry(ZipEntry("manifest.json"))
                zip.write(manifest.toByteArray(Charsets.UTF_8))
                zip.closeEntry()
            }

            Result.success(ExportReport(exported, failed.size, failed))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Imports documents from a recovery bundle zip.
     *
     * For each document:
     * 1. Reads the recovery wrap and unwraps the DEK with the passphrase via Rust FFI
     * 2. Re-wraps the DEK with the local device keystore
     * 3. Stores the .enc file and metadata
     * 4. Zeroizes the DEK
     */
    suspend fun importBundle(
        bundleUri: Uri,
        passphrase: String
    ): Result<ImportReport> = withContext(Dispatchers.IO) {
        val failed = mutableListOf<String>()
        var imported = 0
        var skipped = 0

        try {
            val entries = mutableMapOf<String, MutableMap<String, ByteArray>>()

            // First pass: read all entries from zip
            ZipInputStream(inputStreamProvider(bundleUri)).use { zip ->
                var entry = zip.nextEntry
                while (entry != null) {
                    val name = entry.name
                    val data = zip.readBytes()

                    when {
                        name.startsWith("documents/") && name.endsWith(".enc") -> {
                            val docId = name.removePrefix("documents/").removeSuffix(".enc")
                            entries.getOrPut(docId) { mutableMapOf() }["enc"] = data
                        }
                        name.startsWith("metadata/") && name.endsWith(".meta.json") -> {
                            val docId = name.removePrefix("metadata/").removeSuffix(".meta.json")
                            entries.getOrPut(docId) { mutableMapOf() }["meta"] = data
                        }
                        name.startsWith("wraps/") && name.endsWith(".wraps.json") -> {
                            val docId = name.removePrefix("wraps/").removeSuffix(".wraps.json")
                            entries.getOrPut(docId) { mutableMapOf() }["wraps"] = data
                        }
                    }

                    zip.closeEntry()
                    entry = zip.nextEntry
                }
            }

            // Second pass: import each document
            for ((docId, files) in entries) {
                try {
                    val encData = files["enc"] ?: continue
                    val metaData = files["meta"] ?: continue
                    val wrapsData = files["wraps"] ?: continue

                    // Check for duplicate
                    val existing = when (val result = metadataRepository.get(docId)) {
                        is SecureCoreResult.Success -> result.value
                        is SecureCoreResult.Error -> null
                    }
                    if (existing != null) {
                        skipped++
                        continue
                    }

                    // Parse recovery wrap and unwrap DEK
                    val wrapsJson = String(wrapsData, Charsets.UTF_8)
                    val wrapsObj = JSONObject(wrapsJson)
                    val recoveryObj = wrapsObj.getJSONObject("recovery")
                    val recoveryJson = recoveryObj.toString()

                    val dek = SecureCoreLib.unwrapDekWithPassphrase(recoveryJson, passphrase)
                    try {
                        // Re-wrap with device keystore
                        val deviceWrappedDek = keyManager.wrapDek(dek)

                        // Store encrypted file
                        documentStore.writeDocument(docId, encData)

                        // Parse and save metadata
                        val metaObj = JSONObject(String(metaData, Charsets.UTF_8))
                        val entity = DocumentEntity(
                            docId = metaObj.getString("docId"),
                            filename = metaObj.getString("filename"),
                            mimeType = metaObj.optString("mimeType", null),
                            createdAt = metaObj.getLong("createdAt"),
                            plaintextSize = if (metaObj.has("plaintextSize")) metaObj.getLong("plaintextSize") else null,
                            ciphertextSize = metaObj.getLong("ciphertextSize"),
                            contentHash = metaObj.optString("contentHash", null),
                            wrappedDek = deviceWrappedDek
                        )

                        when (val saveResult = metadataRepository.save(entity)) {
                            is SecureCoreResult.Success -> imported++
                            is SecureCoreResult.Error -> {
                                documentStore.deleteDocument(docId)
                                failed.add(docId)
                            }
                        }
                    } finally {
                        dek.fill(0)
                    }
                } catch (e: Exception) {
                    failed.add(docId)
                }
            }

            Result.success(ImportReport(imported, skipped, failed.size, failed))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // ── Private helpers ─────────────────────────────────────────────

    private suspend fun exportDocument(
        zip: ZipOutputStream,
        docId: String,
        passphrase: String
    ) {
        val entity = when (val result = metadataRepository.get(docId)) {
            is SecureCoreResult.Success -> result.value
                ?: throw IllegalStateException("Document not found: $docId")
            is SecureCoreResult.Error -> throw result.error
        }

        // Unwrap DEK from device keystore
        val dek = keyManager.unwrapDek(entity.wrappedDek)
        try {
            // Re-wrap DEK with passphrase via Rust FFI
            val recoveryWrapJson = SecureCoreLib.wrapDekWithPassphrase(dek, passphrase)

            // Build wraps envelope for bundle
            val wrapsEnvelope = JSONObject().apply {
                put("schema_version", "1.1")
                put("device", JSONObject.NULL)
                put("recovery", JSONObject(recoveryWrapJson))
            }

            // Write encrypted document
            val encData = documentStore.readDocument(docId)
            zip.putNextEntry(ZipEntry("documents/$docId.enc"))
            zip.write(encData)
            zip.closeEntry()

            // Write metadata (without wrappedDek)
            val metaJson = JSONObject().apply {
                put("docId", entity.docId)
                put("filename", entity.filename)
                put("mimeType", entity.mimeType)
                put("createdAt", entity.createdAt)
                put("plaintextSize", entity.plaintextSize)
                put("ciphertextSize", entity.ciphertextSize)
                put("contentHash", entity.contentHash)
            }
            zip.putNextEntry(ZipEntry("metadata/$docId.meta.json"))
            zip.write(metaJson.toString().toByteArray(Charsets.UTF_8))
            zip.closeEntry()

            // Write recovery wrap
            zip.putNextEntry(ZipEntry("wraps/$docId.wraps.json"))
            zip.write(wrapsEnvelope.toString().toByteArray(Charsets.UTF_8))
            zip.closeEntry()
        } finally {
            dek.fill(0)
        }
    }

    private fun buildManifest(exportedDocIds: List<String>): String {
        val sorted = exportedDocIds.sorted()
        val checksumInput = sorted.joinToString("\n")
        val digest = MessageDigest.getInstance("SHA-256")
        val checksum = digest.digest(checksumInput.toByteArray(Charsets.UTF_8))
            .joinToString("") { "%02x".format(it) }

        val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }

        return JSONObject().apply {
            put("format", "recovery_bundle_v1")
            put("version", 1)
            put("created_at", dateFormat.format(Date()))
            put("document_count", sorted.size)
            put("checksum", checksum)
        }.toString(2)
    }
}

File: ios/SecureCore/Sources/SecureCore/Export/ExportService.swift
import Foundation
import ZIPFoundation

/// Errors specific to export/import operations.
public enum ExportServiceError: Error, Equatable {
    case documentNotFound(docId: String)
    case unwrapFailed(String)
    case wrapFailed(String)
    case invalidBundle(String)
    case ioError(String)
}

public struct ExportReport: Equatable {
    public let exportedCount: Int
    public let failedCount: Int
    public let failedDocIds: [String]
}

public struct ImportReport: Equatable {
    public let importedCount: Int
    public let skippedCount: Int
    public let failedCount: Int
    public let failedDocIds: [String]
}

/// Handles export and import of recovery bundles for cloudless document transfer.
///
/// See docs/recovery-format-v1.md for the bundle format specification.
public final class ExportService {
    private let secureCoreLib: SecureCoreLibProtocol
    private let keyManager: KeyManagerProtocol
    private let documentStore: DocumentStoreProtocol
    private let metadataRepository: MetadataRepositoryProtocol

    public init(
        secureCoreLib: SecureCoreLibProtocol,
        keyManager: KeyManagerProtocol,
        documentStore: DocumentStoreProtocol,
        metadataRepository: MetadataRepositoryProtocol
    ) {
        self.secureCoreLib = secureCoreLib
        self.keyManager = keyManager
        self.documentStore = documentStore
        self.metadataRepository = metadataRepository
    }

    // MARK: - Export

    /// Exports selected documents into a recovery bundle zip at the given URL.
    public func exportBundle(
        docIds: [String],
        passphrase: String,
        outputURL: URL
    ) async throws -> ExportReport {
        var exported = 0
        var failedDocIds: [String] = []

        let tempDir = FileManager.default.temporaryDirectory
            .appendingPathComponent(UUID().uuidString)
        try FileManager.default.createDirectory(at: tempDir, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: tempDir) }

        // Create subdirectories
        let docsDir = tempDir.appendingPathComponent("documents")
        let metaDir = tempDir.appendingPathComponent("metadata")
        let wrapsDir = tempDir.appendingPathComponent("wraps")
        try FileManager.default.createDirectory(at: docsDir, withIntermediateDirectories: true)
        try FileManager.default.createDirectory(at: metaDir, withIntermediateDirectories: true)
        try FileManager.default.createDirectory(at: wrapsDir, withIntermediateDirectories: true)

        for docId in docIds {
            do {
                try exportDocument(
                    docId: docId,
                    passphrase: passphrase,
                    docsDir: docsDir,
                    metaDir: metaDir,
                    wrapsDir: wrapsDir
                )
                exported += 1
            } catch {
                failedDocIds.append(docId)
            }
        }

        // Write manifest
        let exportedIds = docIds.filter { !failedDocIds.contains($0) }
        let manifest = buildManifest(docIds: exportedIds)
        let manifestData = try JSONSerialization.data(
            withJSONObject: manifest,
            options: [.prettyPrinted, .sortedKeys]
        )
        try manifestData.write(to: tempDir.appendingPathComponent("manifest.json"))

        // Create zip
        try FileManager.default.zipItem(at: tempDir, to: outputURL)

        return ExportReport(
            exportedCount: exported,
            failedCount: failedDocIds.count,
            failedDocIds: failedDocIds
        )
    }

    // MARK: - Import

    /// Imports documents from a recovery bundle zip.
    public func importBundle(
        bundleURL: URL,
        passphrase: String
    ) async throws -> ImportReport {
        var imported = 0
        var skipped = 0
        var failedDocIds: [String] = []

        let tempDir = FileManager.default.temporaryDirectory
            .appendingPathComponent(UUID().uuidString)
        try FileManager.default.createDirectory(at: tempDir, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: tempDir) }

        // Unzip
        try FileManager.default.unzipItem(at: bundleURL, to: tempDir)

        // Find the content root (may be nested in a subdirectory)
        let contentRoot = try findContentRoot(in: tempDir)

        // Enumerate documents
        let docsDir = contentRoot.appendingPathComponent("documents")
        let metaDir = contentRoot.appendingPathComponent("metadata")
        let wrapsDir = contentRoot.appendingPathComponent("wraps")

        let encFiles = try FileManager.default.contentsOfDirectory(
            at: docsDir,
            includingPropertiesForKeys: nil
        ).filter { $0.pathExtension == "enc" }

        for encFile in encFiles {
            let docId = encFile.deletingPathExtension().lastPathComponent

            do {
                // Check for duplicate
                if let _ = try? metadataRepository.find(docId: docId) {
                    skipped += 1
                    continue
                }

                // Read files
                let encData = try Data(contentsOf: encFile)
                let metaData = try Data(contentsOf: metaDir.appendingPathComponent("\(docId).meta.json"))
                let wrapsData = try Data(contentsOf: wrapsDir.appendingPathComponent("\(docId).wraps.json"))

                // Parse recovery wrap
                guard let wrapsObj = try JSONSerialization.jsonObject(with: wrapsData) as? [String: Any],
                      let recoveryObj = wrapsObj["recovery"] as? [String: Any] else {
                    failedDocIds.append(docId)
                    continue
                }

                let recoveryData = try JSONSerialization.data(withJSONObject: recoveryObj)

                // Unwrap DEK with passphrase via Rust FFI
                var dek = try SecureCoreLibBridge.unwrapDekWithPassphrase(recoveryData, passphrase: passphrase)
                defer { dek.resetBytes(in: dek.startIndex..<dek.endIndex) }

                // Re-wrap with device keychain
                let deviceWrapsJson = try keyManager.wrapDek(dek)

                // Store encrypted file
                try documentStore.writeDocument(docId: docId, data: encData)

                // Parse and save metadata
                guard let metaObj = try JSONSerialization.jsonObject(with: metaData) as? [String: Any] else {
                    _ = try? documentStore.deleteDocument(docId: docId)
                    failedDocIds.append(docId)
                    continue
                }

                let record = DocumentRecord(
                    docId: metaObj["docId"] as? String ?? docId,
                    filename: metaObj["filename"] as? String ?? "unknown",
                    mimeType: metaObj["mimeType"] as? String,
                    createdAt: metaObj["createdAt"] as? Int64 ?? Int64(Date().timeIntervalSince1970 * 1000),
                    plaintextSize: metaObj["plaintextSize"] as? Int64,
                    ciphertextSize: metaObj["ciphertextSize"] as? Int64 ?? Int64(encData.count),
                    wrapsJson: String(data: deviceWrapsJson, encoding: .utf8) ?? ""
                )

                try metadataRepository.save(record)
                imported += 1
            } catch {
                failedDocIds.append(docId)
            }
        }

        return ImportReport(
            importedCount: imported,
            skippedCount: skipped,
            failedCount: failedDocIds.count,
            failedDocIds: failedDocIds
        )
    }

    // MARK: - Private

    private func exportDocument(
        docId: String,
        passphrase: String,
        docsDir: URL,
        metaDir: URL,
        wrapsDir: URL
    ) throws {
        guard let record = try metadataRepository.find(docId: docId) else {
            throw ExportServiceError.documentNotFound(docId: docId)
        }

        guard let wrapsData = record.wrapsJson.data(using: .utf8) else {
            throw ExportServiceError.unwrapFailed("Invalid wrapsJson encoding")
        }

        // Unwrap DEK from device keychain
        var dek = try keyManager.unwrapDek(wrapsData)
        defer { dek.resetBytes(in: dek.startIndex..<dek.endIndex) }

        // Re-wrap DEK with passphrase via Rust FFI
        let recoveryWrapData = try SecureCoreLibBridge.wrapDekWithPassphrase(dek, passphrase: passphrase)

        // Build wraps envelope
        guard let recoveryObj = try JSONSerialization.jsonObject(with: recoveryWrapData) as? [String: Any] else {
            throw ExportServiceError.wrapFailed("Invalid recovery wrap JSON")
        }

        let wrapsEnvelope: [String: Any] = [
            "schema_version": "1.1",
            "device": NSNull(),
            "recovery": recoveryObj
        ]
        let wrapsEnvelopeData = try JSONSerialization.data(withJSONObject: wrapsEnvelope)

        // Write encrypted document
        let encData = try documentStore.readDocument(docId: docId)
        try encData.write(to: docsDir.appendingPathComponent("\(docId).enc"))

        // Write metadata (without wrappedDek)
        let metaDict: [String: Any?] = [
            "docId": record.docId,
            "filename": record.filename,
            "mimeType": record.mimeType,
            "createdAt": record.createdAt,
            "plaintextSize": record.plaintextSize,
            "ciphertextSize": record.ciphertextSize,
            "contentHash": record.contentHash
        ]
        let metaData = try JSONSerialization.data(
            withJSONObject: metaDict.compactMapValues { $0 }
        )
        try metaData.write(to: metaDir.appendingPathComponent("\(docId).meta.json"))

        // Write recovery wrap
        try wrapsEnvelopeData.write(to: wrapsDir.appendingPathComponent("\(docId).wraps.json"))
    }

    private func buildManifest(docIds: [String]) -> [String: Any] {
        let sorted = docIds.sorted()
        let checksumInput = sorted.joined(separator: "\n")
        let checksumData = checksumInput.data(using: .utf8) ?? Data()
        let checksum = sha256Hex(checksumData)

        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]

        return [
            "format": "recovery_bundle_v1",
            "version": 1,
            "created_at": formatter.string(from: Date()),
            "document_count": sorted.count,
            "checksum": checksum
        ]
    }

    private func sha256Hex(_ data: Data) -> String {
        var hash = [UInt8](repeating: 0, count: 32)
        data.withUnsafeBytes { bytes in
            _ = CC_SHA256(bytes.baseAddress, CC_LONG(data.count), &hash)
        }
        return hash.map { String(format: "%02x", $0) }.joined()
    }

    private func findContentRoot(in dir: URL) throws -> URL {
        // Check if manifest.json is directly in dir
        if FileManager.default.fileExists(atPath: dir.appendingPathComponent("manifest.json").path) {
            return dir
        }
        // Check one level deep (zip may contain a root folder)
        let contents = try FileManager.default.contentsOfDirectory(
            at: dir, includingPropertiesForKeys: [.isDirectoryKey]
        )
        for item in contents {
            var isDir: ObjCBool = false
            if FileManager.default.fileExists(atPath: item.path, isDirectory: &isDir), isDir.boolValue {
                if FileManager.default.fileExists(atPath: item.appendingPathComponent("manifest.json").path) {
                    return item
                }
            }
        }
        throw ExportServiceError.invalidBundle("manifest.json not found in bundle")
    }
}

// MARK: - SecureCoreLibBridge extension for recovery operations

/// Bridge to Rust FFI for passphrase-based DEK wrapping/unwrapping.
/// The actual implementation calls the C functions from secure_core.h.
public enum SecureCoreLibBridge {
    /// Wraps a DEK with a passphrase. Returns JSON-encoded RecoveryWrap.
    public static func wrapDekWithPassphrase(_ dek: Data, passphrase: String) throws -> Data {
        // TODO: Call secure_core_wrap_dek_with_passphrase via C bridge
        fatalError("Not yet implemented — requires FFI binding")
    }

    /// Unwraps a DEK from a JSON-encoded RecoveryWrap using the passphrase.
    public static func unwrapDekWithPassphrase(_ recoveryWrapJson: Data, passphrase: String) throws -> Data {
        // TODO: Call secure_core_unwrap_dek_with_passphrase via C bridge
        fatalError("Not yet implemented — requires FFI binding")
    }
}
